src/main/scala/firrtl/passes/InferWidths.scala

```diff
@@ -45,200 +40,251 @@ import firrtl.WrappedExpression._
-           w.arg1 match {
-             case (v:VarWidth) => if (n == v.name) v else w
-             case (v) => w }}
-         case (w) => w
-       }
-       //;println-all-debug(["After removing cycle for " n ", returning " wx])
-       wx
-     }
-     def self_rec (n:String,w:Width) : Boolean = {
-       var has = false
-       def look (w:Width) : Width = {
-         (w map (look)) match {
-           case (w:VarWidth) => if (w.name == n) has = true
-           case (w) => w }
-         w }
-       look(w)
-       has }
-         
-     //; Forward solve
-     //; Returns a solved list where each constraint undergoes:
-     //;  1) Continuous Solving (using triangular solving)
-     //;  2) Remove Cycles
-     //;  3) Move to solved if not self-recursive
-     val u = make_unique(l)
-     
-     //println("======== UNIQUE CONSTRAINTS ========")
-     //for (x <- u) { println(x) }
-     //println("====================================")
-     
-  
-     val f = LinkedHashMap[String,Width]()
-     val o = ArrayBuffer[String]()
-     for (x <- u) {
-       //println("==== SOLUTIONS TABLE ====")
-       //for (x <- f) println(x)
-       //println("=========================")
  
-       val (n, e) = (x._1, x._2)
-       val e_sub = substitute(f)(e)
+    //; Forward solve
+    //; Returns a solved list where each constraint undergoes:
+    //;  1) Continuous Solving (using triangular solving)
+    //;  2) Remove Cycles
+    //;  3) Move to solved if not self-recursive
+    val u = make_unique(l)
+    
+    //println("======== UNIQUE CONSTRAINTS ========")
+    //for (x <- u) { println(x) }
+    //println("====================================")
+ 
+    val f = LinkedHashMap[String, Width]()
+    val o = ArrayBuffer[String]()
+    for ((n, e) <- u) {
+      //println("==== SOLUTIONS TABLE ====")
+      //for (x <- f) println(x)
+      //println("=========================")
 
-       //println("Solving " + n + " => " + e)
-       //println("After Substitute: " + n + " => " + e_sub)
-       //println("==== SOLUTIONS TABLE (Post Substitute) ====")
-       //for (x <- f) println(x)
-       //println("=========================")
+      val e_sub = substitute(f)(e)
 
-       val ex = remove_cycle(n)(e_sub)
+      //println("Solving " + n + " => " + e)
+      //println("After Substitute: " + n + " => " + e_sub)
+      //println("==== SOLUTIONS TABLE (Post Substitute) ====")
+      //for (x <- f) println(x)
+      //println("=========================")
 
-       //println("After Remove Cycle: " + n + " => " + ex)
-       if (!self_rec(n,ex)) {
-         //println("Not rec!: " + n + " => " + ex)
-         //println("Adding [" + n + "=>" + ex + "] to Solutions Table")
-         o += n
-         f(n) = ex
-       }
-     }
-  
-     //println("Forward Solved Constraints")
-     //for (x <- f) println(x)
-  
-     //; Backwards Solve
-     val b = LinkedHashMap[String,Width]()
-     for (i <- 0 until o.size) {
-       val n = o(o.size - 1 - i)
```

Comment:
This iteration order is very important - you must iterate backwards through o. Please change.